### ISSUE TITLE
To set custom URL in the project

### ISSUE THREAD

**Issue by junaidyasin** [upnext/BeaconControl_Android_SDK#13]
I am struggling to know how do we set our own server URL in the Android SDK, there is no clear instruction on how to do so in the integration guide. Here is the part of my logcat, any help would be great!

D/BeaconSDK~BeaconService: Start scan.
D/BeaconSDK~BeaconService: Cannot add client, conf is null.
E/AndroidRuntime: FATAL EXCEPTION: OkHttp Dispatcher Process: com.apps.echnotech.beaconcontrolandroid, PID: 21542                                                                                         java.lang.NullPointerException: Attempt to read from field 'java.lang.String com.upnext.beaconos.sdk.backend.model.TokenResponse.accessToken' on a null object reference at com.upnext.beaconos.sdk.backend.storage.SdkTokenStorage.storeToken(SdkTokenStorage.java:33)

**Comment by roman-upnext** [comment]
@junaidyasin There is a string resource value which you have to overwrite in your application:
`<string name="sdk_config__service_base_url">https://admin.beaconcontrol.io/api/v1/</string>`
Let me know if you have any further questions or problems regarding this setup.

**Comment by junaidyasin** [comment]
Thank you for your prompt response! I have added my custom server and now the application is not crashing, but I am getting these now:

W/BeaconSDK~BeaconControlManagerImpl: Token has not been retrieved.
D/BeaconSDK~HttpCallMediator: IO_ERROR
D/BeaconSDK~BeaconServiceHelper: Error in getConfigurations task, IO_ERROR

I have set the context to getApplicationContext(), and also configured the Client ID and Client Secret right, also I have replaced the userId with my email

**Comment by roman-upnext** [comment]
Could you have a look at issue #12 ?
We're investigating similar issue there and all the HTTP logs that you provide will be very helpful.

**Comment by junaidyasin** [comment]
@roman-upnext Thanks I have already looked into it, unfortunately I cannot provide HTTP logs at this moment, can you please take a look at the complete log file, meanwhile I will try to find a solution to capture HTTP logs as I am on a windows machine

`12-28` 13:06:12.860 2789-2789/com.apps.echnotech.beaconcontrolandroid W/System: ClassLoader referenced unknown path: /data/app/com.apps.echnotech.beaconcontrolandroid-2/lib/arm
12-28 13:06:13.066 2789-2789/com.apps.echnotech.beaconcontrolandroid W/art: Before Android 4.1, method android.graphics.PorterDuffColorFilter android.support.graphics.drawable.VectorDrawableCompat.updateTintFilter(android.graphics.PorterDuffColorFilter, android.content.res.ColorStateList, android.graphics.PorterDuff$Mode) would have incorrectly overridden the package-private method in android.graphics.drawable.Drawable
12-28 13:06:13.640 2789-2789/com.apps.echnotech.beaconcontrolandroid D/BeaconParser: Parsing beacon layout: m:2-3=beac,i:4-19,i:20-21,i:22-23,p:24-24,d:25-25
12-28 13:06:14.141 2789-2789/com.apps.echnotech.beaconcontrolandroid D/BeaconParser: Parsing beacon layout: m:2-3=0215,i:4-19,i:20-21,i:22-23,p:24-24,d:25-25
12-28 13:06:14.153 2789-2789/com.apps.echnotech.beaconcontrolandroid D/BeaconSDK~BeaconService: Start scan.
12-28 13:06:14.153 2789-2789/com.apps.echnotech.beaconcontrolandroid D/BeaconSDK~BeaconService: Cannot add client, conf is null.
12-28 13:06:14.165 2789-2864/com.apps.echnotech.beaconcontrolandroid D/libc: [NET] android_getaddrinfofornetcontext+,hn 22(0x636f6e74726f6c),sn(),hints(known),family 0,flags 1024, proc=com.apps.echnotech.beaconcontrolandroid
12-28 13:06:14.165 2789-2864/com.apps.echnotech.beaconcontrolandroid D/libc: [NET] android_getaddrinfo_proxy+
12-28 13:06:14.166 2789-2864/com.apps.echnotech.beaconcontrolandroid D/libc: [NET] android_getaddrinfo_proxy get netid:0
12-28 13:06:14.187 2789-2865/com.apps.echnotech.beaconcontrolandroid I/Adreno-EGL: <qeglDrvAPI_eglInitialize:379>: EGL 1.4 QUALCOMM build: Nondeterministic_AU_msm8974_LA.BF.1.1.3__release_AU (I3fa967cfef)
                                                                                   OpenGL ES Shader Compiler Version: E031.28.00.02
                                                                                   Build Date: 09/29/15 Tue
                                                                                   Local Branch: mybranch14683032
                                                                                   Remote Branch: quic/master
                                                                                   Local Patches: NONE
                                                                                   Reconstruct Branch: NOTHING
12-28 13:06:14.227 2789-2789/com.apps.echnotech.beaconcontrolandroid W/ModelSpecificDistanceCalculator: Cannot find match for this device.  Using default
12-28 13:06:14.227 2789-2789/com.apps.echnotech.beaconcontrolandroid W/ModelSpecificDistanceCalculator: Cannot find match for this device.  Using default
12-28 13:06:14.303 2789-2864/com.apps.echnotech.beaconcontrolandroid D/libc: [NET] android_getaddrinfo_proxy-, success
12-28 13:06:14.323 2789-2789/com.apps.echnotech.beaconcontrolandroid D/BeaconSDK~BeaconService: onBeaconServiceConnect
12-28 13:06:23.881 2789-2864/com.apps.echnotech.beaconcontrolandroid D/BeaconSDK~BeaconControlManagerImpl: Error in getToken task.
12-28 13:06:23.881 2789-2864/com.apps.echnotech.beaconcontrolandroid W/BeaconSDK~BeaconControlManagerImpl: Token has not been retrieved.
12-28 13:06:23.884 2789-2789/com.apps.echnotech.beaconcontrolandroid D/BeaconSDK~HttpCallMediator: IO_ERROR
12-28 13:06:23.884 2789-2789/com.apps.echnotech.beaconcontrolandroid D/BeaconSDK~BeaconServiceHelper: Error in getConfigurations task, IO_ERROR
12-28 13:06:23.884 2789-2789/com.apps.echnotech.beaconcontrolandroid E/HEY: Some error occured in Beacon SDK: IO_ERROR`

**Comment by roman-upnext** [comment]
Unfortunately there is not enough information in the logs to say exactly what went wrong.
There is for sure a problem when fetching a token (and that is why the getConfigurations call fails with IO_ERROR).
Have you provided the correct `clientId` and `clientSecret` values for your custom backend application instance?

**Comment by junaidyasin** [comment]
Yes I have checked them again and also tested through Postman, I am providing the correct values.

**Comment by roman-upnext** [comment]
Then I think we'll need the HTTP logs to be able to help you further.

**Comment by junaidyasin** [comment]
@roman-upnext Thank you good sir for pointing me to the right direction. I managed to sniff the requests and after looking at it for quite a bit. I realized that if I uninstall the app and execute a fresh install every time, then everything goes okay and I get the `/api/v1/configurations` response after receiving the token. Although if I run the app again it skips the part of requesting a new token and don't make any calls to `api/v1/oauth/token`. I think it has something to do with refreshing tokens or requesting new tokens?
Anyway my code in now working and scanning beacons without any issue. Thumbs up for the upnext team.
